### PR TITLE
feat(onboarding): Add a skipVerification flag that can skip phone verification during onboarding

### PR DESCRIPTION
**Pull request by jh2oman**
### Description

This PR adds a feature flag, `skipVerification`, that can be used to toggle whether or not Phone Verification should be a part of the onboarding flow. Removing phone verification from the onboarding flow will be part of an upcoming experiment.

[Experiment Proposal
](https://www.notion.so/valora-inc/Verification-c1afb404d23242d293bd80d2d40c8efb)

[Zenhub Ticket
](https://app.zenhub.com/workspaces/acquisition-squad-sprint-board-6010683afabec1001a090887/issues/valora-inc/wallet/1838)

https://user-images.githubusercontent.com/8432644/157553203-ecab871b-badf-4134-92b7-06f7610635c8.mp4

### Tested

Tested locally on an iOS emulator

### How others should test

Turn `skipVerification` feature flag to `true`. Then go through the onboarding flow (both create account, and restore account steps) and note that phone verification is skipped and the `Step X of X` header displays the correct numbers.

### CODE REVIEW

File: packages/mobile/src/app/selectors.ts [hunk 1]
```diff
@@ -180,29 +182,48 @@ export const registrationStepsSelector = createSelector(
...
+        step =
+          step > (steps as Record<Screens.EnableBiometry, number>)[Screens.EnableBiometry]
+            ? step - 1
+            : step
       }
-      // remove biometry screen from step
-      const step = restoreAccountSteps[activeScreen as RestoreAccountScreens]
-      return { step: step > 3 ? step - 1 : step, totalSteps: 4 }
+    }
+    if (skipVerification) {
```
Comment (cajubelt) [inline packages/mobile/src/app/selectors.ts:220]:
```suggestion
    if (skipVerification && totalSteps > 0) {
```
this is pretty dumb but feels a little safer

File: packages/mobile/src/app/selectors.ts [hunk 2]
```diff
@@ -180,29 +182,48 @@ export const registrationStepsSelector = createSelector(
...
+      totalSteps--
     }
 
-    if (biometryEnabled) {
-      return { step: createAccountSteps[activeScreen as CreateAccountScreens], totalSteps: 4 }
+    return {
+      step,
```
Comment (cajubelt) [inline packages/mobile/src/app/selectors.ts:225]: a hidden assumption here is that phone verification always takes place last in the list of steps. What if it didn't? Shouldn't we make sure `step` is updated to account for that? Could do this with something like `if (step >= steps[Screens.VerificationEducationScreen] || step >= steps[Screens.VerificationInputScreen]) step--`

File: packages/mobile/src/app/selectors.ts [hunk 3]
```diff
@@ -180,29 +182,48 @@ export const registrationStepsSelector = createSelector(
...
+      totalSteps,
     }
-    // remove biometry screen from step
-    const step = createAccountSteps[activeScreen as CreateAccountScreens]
-    return { step: step > 3 ? step - 1 : step, totalSteps: 3 }
```
Comment (MuckT) [inline packages/mobile/src/app/selectors.ts:206]: I noticed previously the code would return out of the if blocks when conditions were hit. Here it looks like here we are switching to modifying the `totalSteps`, `step` and `steps` across the if blocks as there are now multiple conditions; e.g. I could have !biometryEnabled and also be in the experiment group so `totalSteps`, `step` and `steps` so it needs to dynamically account for those two conditions; Is this correct? I don't doubt that this works, but it's a bit hard to follow.

File: packages/mobile/src/import/saga.test.ts [hunk 4]
```diff
@@ -79,6 +84,29 @@ describe('Import wallet saga', () => {
     await expectSuccessfulSagaWithPhrase(mockValidSpanishPhrase)
   })
 
+  it('initializes account and navigates to home if skipVerification is true', async () => {
```
Comment (cajubelt) [inline packages/mobile/src/import/saga.test.ts:87]: is there also coverage on the case where skipVerification is false (ensuring that the verification screens are shown)?

File: packages/mobile/src/app/selectors.ts (hunk 3 above)
Comment (jh2oman) [reply packages/mobile/src/app/selectors.ts:206]: Yes thats correct, it needs to be able to account for the possibility of both the biometrics screen and the phone verification screens being skipped.

The code looks much more confusing in diff mode than it does in split diff or on its own.

File: packages/mobile/src/redux/migrations.test.ts [hunk 5]
```diff
@@ -428,6 +429,15 @@ describe('Redux persist migrations', () => {
...
     expectedSchema.app.showPriceChangeIndicatorInBalances = false
 
+    expect(migratedSchema).toMatchObject(expectedSchema)
+  })
+  it('works for v37 to v38', () => {
+    const oldSchema = v37Schema
+    const migratedSchema = migrations[38](oldSchema)
+
+    const expectedSchema: any = { ...oldSchema }
+    expectedSchema.app.skipVerification = false
```
Comment (cajubelt) [inline packages/mobile/src/redux/migrations.test.ts:439]: we should probably use a deep copy here

Comment (cajubelt) [reply packages/mobile/src/redux/migrations.test.ts:439]: since this mutates the oldSchema object

File: packages/mobile/src/import/saga.ts [hunk 6]
```diff
@@ -160,8 +161,13 @@ export function* importBackupPhraseSaga({ phrase, useEmptyWallet }: ImportBackup
       ValoraAnalytics.track(AppEvents.redux_store_recovery_success, { account })
     }
     ValoraAnalytics.track(OnboardingEvents.wallet_import_success)
-
-    navigateClearingStack(Screens.VerificationEducationScreen)
+    const skipVerification = yield select(skipVerificationSelector)
+    if (skipVerification) {
+      yield put(initializeAccount())
```
Comment (cajubelt) [inline packages/mobile/src/import/saga.ts:166]: seems like there's probably a good answer to this, but can you explain to me why we only need to initialize the account if we're skipping verification?

**Review by cajubelt — COMMENTED** [review]
minor stuff, looks good overall

File: packages/mobile/src/redux/migrations.test.ts (hunk 5 above)
Comment (jh2oman) [reply packages/mobile/src/redux/migrations.test.ts:439]: 🤯

File: packages/mobile/src/app/selectors.ts (hunk 2 above)
Comment (jh2oman) [reply packages/mobile/src/app/selectors.ts:225]: That would help safeguard against future changes. I can add it in.

File: packages/mobile/src/import/saga.test.ts (hunk 4 above)
Comment (jh2oman) [reply packages/mobile/src/import/saga.test.ts:87]: Adding it in 👍

File: packages/mobile/src/import/saga.ts (hunk 6 above)
Comment (jh2oman) [reply packages/mobile/src/import/saga.ts:166]: Currently `initializeAccount` is only called on the VerificationEducationScreen OR if it gets skipped because komenci is down or verification is disabled completely. I'm not entirely sure why those were the places chosen, but I though it probably best to continue calling `initializeAccount` when verification is skipped

File: packages/mobile/src/app/selectors.ts [hunk 7]
```diff
@@ -218,7 +218,16 @@ export const registrationStepsSelector = createSelector(
       }
     }
     if (skipVerification) {
-      totalSteps--
+      if (Object.keys(steps).includes(Screens.VerificationEducationScreen)) {
+        totalSteps--
+        step =
+          step >
```
Comment (cajubelt) [inline packages/mobile/src/app/selectors.ts:224]:
```suggestion
          step >=
```
I think this needs to be >=

Comment (cajubelt) [reply packages/mobile/src/app/selectors.ts:224]: not 100% sure

**Review by cajubelt — APPROVED** [review]
nice! see comment, not sure if we need it to be >= in one place

File: packages/mobile/src/app/selectors.ts (hunk 7 above)
Comment (jh2oman) [reply packages/mobile/src/app/selectors.ts:224]: Equal to would be the actual VerificationEducationScreen, which should hopefully never be shown when skipVerification is true

File: packages/mobile/src/app/selectors.ts [hunk 8]
```diff
@@ -180,29 +182,57 @@ export const registrationStepsSelector = createSelector(
...
+    let steps
+    let totalSteps
+    let step
     if (recoveringFromStoreWipe) {
-      return {
-        step: storeWipeRecoverySteps[activeScreen as StoreWipeRecoveryScreens],
-        totalSteps: 3,
-      }
+      steps = storeWipeRecoverySteps
+      totalSteps = 3
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply packages/mobile/src/app/selectors.ts:199]: It does happens to be the same since verification has 2 screens in the same step. I was hesitant to code that directly in b/c i'm not sure how stable that will be over time

File: packages/mobile/src/app/selectors.test.ts [hunk 9]
```diff
@@ -129,6 +129,22 @@ describe('registrationStepsSelector', () => {
       }
     )
   })
+  it('should return the correct steps for create account screens with skipValidation enabled', () => {
+    ;(Object.keys(createAccountSteps) as Array<keyof typeof createAccountSteps>).forEach(
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply packages/mobile/src/app/selectors.test.ts:133]: I don't think it actually has any purpose. I'll admit that i copy/pasted from the other tests which had this pattern

**Comment by ValoraQA** [comment]
Hey @MuckT please help us to test the above PR as it has some feature to be enabled from backend
